### PR TITLE
fix ofa search space when model has shortcut

### PR DESCRIPTION
**Pull request by ceci3**
fix ofa search space when model has shortcut
fix export sub-conv param
move BaseBlock to common file 
fix dataparallel
change  name of search space from block.key to layers' name
change export functional

### CODE REVIEW

File: paddleslim/nas/ofa/get_sub_model.py [hunk 1]
```diff
@@ -68,40 +75,107 @@ def get_prune_params_config(graph, origin_model_config):
...
+    same_ss_per_op = []
+    pre_ops = graph.pre_ops(op)
+    for pre_op in pre_ops:
+        if pre_op.type() in WEIGHT_OP:
+            return
+        same_ss_per_op = _find_weight_ops(pre_op, graph, same_ss_per_op)
+    return same_ss_per_op
+
+
+def check_ss(graph):
```
Comment (qingqing01) [inline paddleslim/nas/ofa/get_sub_model.py]: ss -> search_space,   need more clear name  and add comments

File: paddleslim/nas/ofa/get_sub_model.py [hunk 2]
```diff
@@ -68,40 +75,107 @@ def get_prune_params_config(graph, origin_model_config):
...
+
+
+def _find_weight_ops(op, graph, weights):
+    pre_ops = graph.pre_ops(op)
+    for pre_op in pre_ops:
+        if pre_op.type() in WEIGHT_OP:
+            for inp in pre_op.all_inputs():
+                if inp._var.persistable:
+                    weights.append(inp._var.name)
+            return weights
```
Comment (qingqing01) [inline paddleslim/nas/ofa/get_sub_model.py:150]: find ops or find weithgs ?

File: paddleslim/nas/ofa/get_sub_model.py (hunk 1 above)
Comment (ceci3) [reply paddleslim/nas/ofa/get_sub_model.py]: done, thanks

File: paddleslim/nas/ofa/get_sub_model.py (hunk 2 above)
Comment (ceci3) [reply paddleslim/nas/ofa/get_sub_model.py:150]: find ops with weights

File: paddleslim/nas/ofa/utils/utils.py [hunk 3]
```diff
@@ -62,13 +62,12 @@ def set_state_dict(model, state_dict):
 def remove_model_fn(model, sd):
```
Comment (wanghaoshuang) [inline paddleslim/nas/ofa/utils/utils.py]: sd是state_dict? 其实可以直接用全拼

File: paddleslim/nas/ofa/get_sub_model.py [hunk 4]
```diff
@@ -68,40 +77,115 @@ def get_prune_params_config(graph, origin_model_config):
...
-                    value.shape[1] * out_exp)
-                prune_value = super_value[:in_chn, :out_chn, ...] \
-                                 if super_model_sd != None else value[:in_chn, :out_chn, ...]
+    """ Prune parameters according to the config.
+    """
+    for l_name, sublayer in model.named_sublayers():
+        if isinstance(sublayer, BaseBlock):
+            continue
+        for p_name, param in sublayer.named_parameters(include_sublayers=False):
+            name = l_name + '.' + p_name
```
Comment (wanghaoshuang) [inline paddleslim/nas/ofa/get_sub_model.py]: name好像只在`super_model_sd != None`时用到了

File: paddleslim/nas/ofa/get_sub_model.py [hunk 5]
```diff
@@ -68,40 +77,115 @@ def get_prune_params_config(graph, origin_model_config):
 
 
 def prune_params(model, param_config, super_model_sd=None):
```
Comment (wanghaoshuang) [inline paddleslim/nas/ofa/get_sub_model.py:79]: 加注释重点说明下`super_model_sd`

File: paddleslim/nas/ofa/get_sub_model.py (hunk 2 above)
Comment (wanghaoshuang) [reply paddleslim/nas/ofa/get_sub_model.py:150]: 找到当前operator之前所有operator的weights?

File: paddleslim/nas/ofa/get_sub_model.py [hunk 6]
```diff
@@ -68,40 +77,115 @@ def get_prune_params_config(graph, origin_model_config):
...
+                if inp._var.persistable:
+                    weights.append(inp._var.name)
+            return weights
+        return _find_weight_ops(pre_op, graph, weights)
+
+
+def _find_pre_elementwise_add(op, graph):
+    """ Find precedor of the elementwise_add operator.
+    """
+    same_ss_per_op = []
```
Comment (wanghaoshuang) [inline paddleslim/nas/ofa/get_sub_model.py]: 变量命名不要用这种看不出意义的缩写

**Review by wanghaoshuang — COMMENTED** [review]
这个PR内容与剪枝功能有重复，可以先修改合并，然后将OFA里的剪枝需求提个issue，由负责剪枝的同学完善一下。

File: paddleslim/nas/ofa/get_sub_model.py (hunk 2 above)
Comment (ceci3) [reply paddleslim/nas/ofa/get_sub_model.py:150]: 找到当前op之前的含有参数的op的参数名字。比如 conv->relu->elementwise_add，输入为elementwise_add，需要查找的是conv的weight 和bias的两个参数

File: paddleslim/nas/ofa/get_sub_model.py (hunk 4 above)
Comment (ceci3) [reply paddleslim/nas/ofa/get_sub_model.py]: done

File: paddleslim/nas/ofa/utils/utils.py (hunk 3 above)
Comment (ceci3) [reply paddleslim/nas/ofa/utils/utils.py]: done， thanks

File: paddleslim/nas/ofa/get_sub_model.py (hunk 5 above)
Comment (ceci3) [reply paddleslim/nas/ofa/get_sub_model.py:79]: done，thanks

File: paddleslim/nas/ofa/get_sub_model.py (hunk 6 above)
Comment (ceci3) [reply paddleslim/nas/ofa/get_sub_model.py]: done，thanks

**Review by wanghaoshuang — APPROVED** [review]

**Pull request merged**